### PR TITLE
Adimensionality + bug fixes

### LINKED ISSUE #216: Failing in compiling with DMM due to extrapolation of detuning limits
### Short description

There are failing jobs due to a failure in DMM being in the limits of the device for some instances. Example:
`Failed to compile the sequence due to: For a detuning map with a total summed weight of 4.111802592873573, the total applied detuning from a DMM pulse with minimum detuning -54.279604 rad/µs goes below the total bottom detuning of the DMM (-125.66370614359172 rad/µs). To respect this constraint, keep the detuning above -30.561707014190684 rad/µs.`

### What is the expected result?

That using DMM do not violate the constraints of the device, regardless of the instance. 

### What is the actual result?

It is violating the constraints of the device, preventing the qubo solving. 

### Steps/Code to reproduce

Used the solverbenchmarker to run experiments
Chose small qubos (sizes 5 - 7), only positive off-diagonal terms and negative diagonals, bounds [-50, 50]
Default configuration + DMM (optimized drive/blade embedding also give this problem).

### Tracebacks (optional)

_No response_

### Environment details (optional)

QuboSolver version 0.8.0

### Would you like to work on this issue?

_No response_

### PR DESCRIPTION
**Pull request by CdeTerra**
Makes the embedding and drive-shaping pipeline fully adimensional so the register and the pulse share a single unit system, and fixes the scaling bugs that followed from the previous mix of adimensional / physical units.

Changes:
- Greedy embedder: adimensional interactions & layout.
- Config: replace `greedy_spacing` / `min_distance` with `greedy_max_possible_term` and `max_min_dist_ratio` for the device.
- Heuristic drive shaper aligned with the register: interaction / detuning is consistent with the QUBO's quadratic/linear ratio (and identical across embedders).
- Clip DMM usage to the device's limit `total_bottom_detuning`.
- Add regression tests covering the adimensional register, register/drive normalization consistency, no-overflow, and quality of solution in additional cases.
- Many other improvements and fixes.

Fixes #216, #185, #198, #199, #200, #201.

### CODE REVIEW

File: qubosolver/algorithms/greedy/greedy.py [hunk 1]
```diff
@@ -181,14 +180,12 @@ def greedy_algorithm(
         results: dict,
         params: dict,
         on_step: Optional[Callable[[Dict[str, Any]], None]] = None,
+        max_radial_distance: float | None = None,
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/greedy/greedy.py]: For such parameters, `float("inf")` is often the best default (instead of `None`).
If I understand correctly, here `None` means "takes the device's" ? If so, I'd rather use
```python
max_radial_distance: float | Literal["device"] = "device"
```
It makes it more explicit, and more scalable as a bonus as we could add other `Literal`s

File: qubosolver/pipeline/program.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    program = QuantumProgram(
+        register=embedding,
+        drive=drive,
+    )
+
+    if device.specs["max_duration"] is None:
+        device_with_duration = dataclasses.replace(
+            device._device,
+            max_sequence_duration=config.drive_shaping.default_sequence_duration,
+        )
```
Comment (abussy-pasqal) [inline qubosolver/pipeline/program.py:30]: kind of hacky. shouldn't we open an issue in qoolqit ?

File: qubosolver/config.py [hunk 3]
```diff
@@ -448,6 +458,13 @@ class SolverConfig(Config):
 
     def __repr__(self) -> str:
         return self.config_name
+    
+    @property
+    def max_min_dist_ratio(self) -> float | None:
```
Comment (abussy-pasqal) [inline qubosolver/config.py]: If I read the docstring correctly, shouldn't it return `self.embedding.max_min_dist_ratio` if it's not `None` ?

Moreover, we should consider a value that matches the former default value of 7um (https://github.com/pasqal-io/qubo-solver/blob/791a5189f36707106514f4898c39d078071e8c44/qubosolver/config.py#L185)
More precisely the `max_radial_distance` of a chosen pulser device, over the spacing of 7um

See the formula you used in the UTs:
```
max_min_dist_ratio = max_min_dist_ratio * device.min_atom_distance / spacing
```

Comment (CdeTerra) [reply qubosolver/config.py]: `greedy_spacing` was meant to be removed, and I've finished doing it now.

File: qubosolver/algorithms/greedy/greedy.py (hunk 1 above)
Comment (CdeTerra) [reply qubosolver/algorithms/greedy/greedy.py]: Agreed, I've added the `'device'` value.

File: qubosolver/pipeline/program.py (hunk 2 above)
Comment (CdeTerra) [reply qubosolver/pipeline/program.py:30]: Yes, but let us manage it as a next task.

File: qubosolver/algorithms/decompose.py [hunk 4]
```diff
@@ -23,6 +25,16 @@
 )
 
 
+def compute_max_min_distances(
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/decompose.py]:
```suggestion
def compute_min_max_distances(
```

File: qubosolver/config.py [hunk 5]
```diff
@@ -183,29 +186,31 @@ class EmbeddingConfig(Config):
             Defaults to `False`.
         animation_save_path (str | None, optional): If provided, path to save animation.
             Defaults to None.
-        min_distance (float | None): Minimum atom separation (μm).
-            If not None, the resulting register will be normalized so that the minimum atom separation is equal to this value. Should be 1.001 when using the Heuristic Drive-Shaping, and None when using the Optimized Drive-Shaping. Defaults to 1.001.
+        max_min_dist_ratio: (float | None): Maximum ratio between the
```
Comment (abussy-pasqal) [inline qubosolver/config.py]: change typing to match implementation

File: qubosolver/algorithms/greedy/layout.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    match layout_type:
+        case LayoutType.TRIANGULAR | "triangular":
+            layout = LayoutType.TRIANGULAR.value(n_traps, spacing=1)
+            return torch.tensor(layout.coords)
+
+        case LayoutType.SQUARE | "square":
+            n = int(torch.ceil(torch.sqrt(2 * torch.tensor(n_traps))).item())
+            layout = LayoutType.SQUARE.value(n, n, spacing=1)
+            coords = torch.tensor(layout.coords)
+            squared_distances = (coords**2).sum(dim=1)
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/greedy/layout.py]:
```suggestion
            squared_distances = coords.square().sum(dim=1)
```

File: qubosolver/algorithms/greedy/greedy.py [hunk 7]
```diff
@@ -176,26 +162,24 @@ def greedy_algorithm(
...
         state snapshot after each placement (and an initial snapshot).
         """
-        max_radial_distance = params.get(
-            "max_radial_distance", params["device"].max_radial_distance
-        )
         nodes = list(range(Q.shape[0]))
 
         vertices = set(nodes)
-        all_traps = set(list(layout.traps_dict.keys()))
+        n_traps = len(coords)
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/greedy/greedy.py:178]: if `coords` is only needed for `n_traps`, why not replacing the former with the latter as an input argument ?

File: qubosolver/config.py [hunk 8]
```diff
@@ -172,8 +171,12 @@ class EmbeddingConfig(Config):
...
             Defaults to ``-1``, i.e. automatically set to match the selected device capacity.
             A too high value will impede computational efficiency.
-        greedy_spacing (float, optional): The minimum distance between atoms.
-            Defaults to 7 (μm).
+        greedy_max_possible_term (float | tuple[Literal['factor'], float]):
+            If a float, it corresponds to the maximum representable quadratic
+            term. If a tuple, the first element should be 'factor', and the
+            second element is a multiplier on the QUBO's maximum quadratic term
+            to define the maximum representable quadratic term.
+            Defaults to ('factor', 1.0).
```
Comment (abussy-pasqal) [inline qubosolver/config.py]: I find the name and doc of this parameters misleading. It leads to believe that it _only_ impacts the maximum quadratic term, whereas it impacts all of them through spacing in a non-trivial way (1/r^6). I can't think of a better way to parametrize `spacing` though. Maybe you could add comments on the subject. And also about how non-trivial and instance-dependent finding an optimal (or good enough) value is (unless I'm mistaken).

File: qubosolver/config.py (hunk 5 above)
Comment (CdeTerra) [reply qubosolver/config.py]: Done

File: qubosolver/algorithms/decompose.py (hunk 4 above)
Comment (CdeTerra) [reply qubosolver/algorithms/decompose.py]: Done

File: qubosolver/algorithms/greedy/layout.py (hunk 6 above)
Comment (CdeTerra) [reply qubosolver/algorithms/greedy/layout.py]: Ok, not sure why this should be better but done anyway.

File: qubosolver/algorithms/greedy/greedy.py (hunk 7 above)
Comment (CdeTerra) [reply qubosolver/algorithms/greedy/greedy.py:178]: Actually, there are many more things to improve on the legacy code, but this should be done separately. The main goal of this PR is on the adimensionality and the functional aspects.

File: qubosolver/config.py (hunk 8 above)
Comment (CdeTerra) [reply qubosolver/config.py]: I have added explanations.

File: tests/test_driveshaping.py [hunk 9]
```diff
@@ -204,11 +210,70 @@ def test_generate_heuristic_drive_shaper(
...
-    check.almost_equal(drive.detuning.min(), -3.0, abs=1e-4)
+    check.almost_equal(drive.detuning.min(), -1.5, abs=1e-4)
     if dmm:
-        check.almost_equal(drive.detuning.max(), 3.0, abs=1e-4)
+        check.almost_equal(drive.detuning.max(), 1.5, abs=1e-4)
     else:
-        check.almost_equal(drive.detuning.max(), 2.0, abs=1e-4)
+        check.almost_equal(drive.detuning.max(), 1.0, abs=1e-4)
+
+
```
Comment (abussy-pasqal) [inline tests/test_driveshaping.py:222]:
```suggestion
@pytest.mark.usefixtures("restore_rng_state")
```

File: tests/solvers/test_solver.py [hunk 10]
```diff
@@ -345,3 +348,66 @@ def test_save_load_qubo_solver_quantum(
...
+        match="DMM final detuning would exceed the device's total_bottom_detuning",
+    ):
+        solution = QuboSolver(instance, config).solve()
+
+    assert solution.bitstrings.numel() > 0
+
+
+def _triangular_register_qubo() -> np.ndarray:
+    data_graph = DataGraph.triangular(4, 4, 1)
+    np.random.seed(0)
```
Comment (abussy-pasqal) [inline tests/solvers/test_solver.py:376]: use the fixture 
```python
@pytest.mark.usefixtures("restore_rng_state")
```
in tests where a seed is set

File: tests/test_driveshaping.py (hunk 9 above)
Comment (CdeTerra) [reply tests/test_driveshaping.py:222]: Done

File: tests/solvers/test_solver.py (hunk 10 above)
Comment (CdeTerra) [reply tests/solvers/test_solver.py:376]: Done

File: qubosolver/pipeline/drive.py [hunk 11]
```diff
@@ -564,21 +566,28 @@ def build_drive(self, params: list) -> Drive:
...
         amp_params = [p * max_amplitude for p in amp_params]
-        det_params = [p * max_detuning for p in det_params]
-
         amp_wave = InterpolatedWaveform(max_seq_duration, amp_params)
+
+        # QoolQit rescales only based on the amplitude, so the maximum
+        # of the detuning depends on the amplitude.
+        det_ratio = max_detuning / max_amplitude
+        det_scale = det_ratio * float(amp_wave.max()) * (1.0 - 1e-3)
+        det_params = [p * det_scale for p in params[3:]]
```
Comment (abussy-pasqal) [inline qubosolver/pipeline/drive.py:634]: This changes the cost function too much. I already proposed this change, and we decided not too implement it with @yassine-naghmouchi

File: qubosolver/algorithms/decompose.py [hunk 12]
```diff
@@ -61,27 +73,24 @@ def __init__(self, id: int, x: float, y: float, weight: float, device: PulserBas
 
 
 def compute_distance_interaction_matrix(
-    device: PulserBaseDevice,
     qubo_matrix: torch.Tensor,
-    neglecting_inter_distance: float = 15.0,
+    neglecting_inter_distance: float = 1.5,
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/decompose.py:74]: why change this value ? I assume the explanation will apply to all distance parameters ?

File: qubosolver/algorithms/decompose.py [hunk 13]
```diff
@@ -143,12 +155,12 @@ def vertices_to_place(
         blocking_vertices = torch.argwhere(torch.eq(distances, 0))
         blocking_vertices = blocking_vertices[blocking_vertices != i]
 
-        # should be >= separation_threshold
+        # equality because the dist matrix has setting on neglecting_inter_distance = 15
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/decompose.py:155]: I agree that equality is enough given the default parameters. What happens if `neglecting_inter_distance != separation_threshold` ? Using `>=`  is more generic as it doesn't assume any pre-condiction on `dist_matrix`.

Similarily, what happens when the distance is `1e-3` ?

IMO this function is too coupled with `compute_distance_interaction_matrix` and some rewriting is needed. But this is another task.

File: qubosolver/algorithms/decompose.py [hunk 14]
```diff
@@ -416,31 +428,34 @@ def cost_interaction_point_continuous(
...
+            each point and the origin.
 
     Returns:
         float: Cost evaluation.
     """
+
     cost = 0.0
-    epsilon = 1e-6
-    interaction_coeff = device.interaction_coeff
+    epsilon = 1e-8 * AnalogDeviceWithDMM()._device.interaction_coeff ** (-1 / 3)
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/decompose.py]: add a comment about the choice of `AnalogeDeviceWithDMM`

File: qubosolver/algorithms/decompose.py [hunk 15]
```diff
@@ -450,6 +465,15 @@ def cost_interaction_point_continuous(
...
             cost += diff**2
 
+    penalty_coeff = sum([coeff**2 for coeff in Q_target])
+    if np.linalg.norm(pos_new) > max_radial_distance:
+        cost += float(penalty_coeff * (1 + np.linalg.norm(pos_new) - max_radial_distance))
+
+    for placed_point in placed_points:
+        dist = np.linalg.norm(np.array(pos_new) - np.array(placed_point))
+        if dist < min_distance:
+            cost += float(penalty_coeff * (1 + min_distance - dist))
```
Comment (abussy-pasqal) [inline qubosolver/algorithms/decompose.py:470]: Can you explain these two added penalties ? Especially, why aren't they infinite ?

File: tests/test_driveshaping.py [hunk 16]
```diff
@@ -204,11 +210,72 @@ def test_generate_heuristic_drive_shaper(
...
+    # greedy labels qubits "q0"/"q1", blade labels them 0/1
+    key = ("q0", "q1") if ("q0", "q1") in interactions else (0, 1)
+    interaction = interactions[key]
+    detuning_wf = inner._drive.detuning
+    detuning = detuning_wf(detuning_wf.duration)
+    return float(interaction / detuning)
+
+
+@pytest.mark.parametrize("embedding_method", ["greedy", "blade"])
+def test_heuristic_register_and_drive_share_normalization(
```
Comment (abussy-pasqal) [inline tests/test_driveshaping.py]:
```suggestion
def test_heuristic_register_and_drive_shape_normalization(
```

File: qubosolver/pipeline/drive.py (hunk 11 above)
Comment (CdeTerra) [reply qubosolver/pipeline/drive.py:634]: After discussions, this change is accepted.

File: qubosolver/algorithms/decompose.py (hunk 14 above)
Comment (CdeTerra) [reply qubosolver/algorithms/decompose.py]: Removed

File: tests/test_driveshaping.py (hunk 16 above)
Comment (CdeTerra) [reply tests/test_driveshaping.py]: Done

**Comment by abussy-pasqal** [comment]
Rebased on v1 refactoring in #223
